### PR TITLE
Allow forms to be rendered with non-post methods

### PR DESCRIPTION
**Pull request by JRaspass**
I'd like to be able to use `Cro::WebApp::Form` for an advanced search route but the render sub is currently hardcoded to do a POST request, this PR changes that.

I'm not sure if this is the right way to go about this, but it seems like the simplest approach.

### CODE REVIEW

**Comment by lizmat** [comment]
Seems like a valid approach to me :-)

**Comment by vendethiel** [comment]
I think we need to forbid file inputs outside of post

**Comment by jnthn** [comment]
:+1: on the feature and API; agree with @vendethiel that we should also raise an error if there's a file control on the form.

**Comment by jnthn** [comment]
(Probably implemented by passing the method along to `HTML-RENDER-DATA`)

**Comment by JRaspass** [comment]
@jnthn That makes sense but I'm also wondering if we ought to avoid adding the `__CSRF_TOKEN` too, which means this ought to happen at a layer above the template? Perhaps the form ought to be annotated as being a `GET` form with a role or something. Thoughts?

**Comment by jnthn** [comment]
@JRaspass If we pass the method along to `HTML-RENDER-DATA` then it would also be able to decide to insert or not insert the CSRF token using it (with a little care over the caching - perhaps storing the method the cache applies to and only using it if it matches; it'd be unusual to render the same form with two different methods).

**Comment by JRaspass** [comment]
Okay silly question, how does one pass arguments to methods in a template?

Works:
```
<&FOO(:$method)>
```

Doesn't work:
```
<.HTML-RENDER-DATA(:$method)>
```

**Comment by jnthn** [comment]
@JRaspass I'd expect that to work, but since the original was `<@HTML-RENDER-DATA.controls>` then it'd need to become something like `<@.HTML-RENDER-DATA(:$method).controls>` or `<@$_.HTML-RENDER-DATA(:$method).controls>` to retrain the iteration.

**Comment by JRaspass** [comment]
@jnthn I was going to pass it to all the calls to `HTML-RENDER-DATA` in the form so the caching dtrt but when I try the above syntax I get the following error:
```
  Died with the exception:
      Template parse failed: malformed topic tag at line 43 near '(:$method).enctype>"</?>>
```
From the following change:
```
enctype="<.HTML-RENDER-DATA(:$method).enctype>"
```
Since I think I ought to make enctype return nil under GET too. I think this isn't the first time I've ran into the inability to pass args to methods from template land, so unless I'm missing something obvious I don't think it's currently possible.

**Comment by jnthn** [comment]
@JRaspass Indeed, I'd forgotten we can't do that. Well, couldn't; I've added support for it in e92617ea6b.

**Comment by JRaspass** [comment]
@jnthn Oh awesome, that'll actually be useful to me in other places, like formatting datetimes in templates. Many thanks. I'll let you know what roadblock I run into next :joy:

**Comment by JRaspass** [comment]
@jnthn I've updated this PR based on your feedback, still not wild about threading the method through all these callsites as opposed to making it a property of the form but it does at least work. Feel free to suggest better wording for the exception too. And finally there's now a small test covering this change :tada:

**Comment by jnthn** [comment]
> still not wild about threading the method through all these callsites as opposed to making it a property of the form but it does at least work

With this we can re-use the same form as both `GET` and `POST`; if we make it a property of the form itself then that's not possible.

File: lib/Cro/WebApp/Form.pm6 [hunk 1]
```diff
@@ -385,6 +390,8 @@ role Cro::WebApp::Form {
         }
         for self.^attributes.grep(*.has_accessor) -> Attribute $attr {
             my ($control-type, %properties) = self!calculate-control-type($attr);
+            die "Can't have type 'file' in a GET form"
+                if $control-type eq 'file' && $method eq 'get';
```
Comment (jnthn) [inline lib/Cro/WebApp/Form.pm6:394]: Slight preference for a typed exception, and also include the class name of the form perhaps.

File: t/form-get.t [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+constant template = parse-template ｢<&form($_, :method('get'))>｣;
+
+with template.render: Search.empty {
+    like $_, / 'method="get"' /, 'Form method is GET';
+
+    unlike $_, / '__CSRF_TOKEN' /, 'CSRF token is omitted';
+}
+
+throws-like { template.render: Upload.empty }, Exception,
+    :message("Can't have type 'file' in a GET form"),
```
Comment (jnthn) [inline t/form-get.t:17]: The test would be more robust with a typed exception too

File: lib/Cro/WebApp/Form.pm6 (hunk 1 above)
Comment (JRaspass) [reply lib/Cro/WebApp/Form.pm6]: done

File: t/form-get.t (hunk 2 above)
Comment (JRaspass) [reply t/form-get.t]: done

**Comment by JRaspass** [comment]
@jnthn Updated with a typed exception.

**Comment by jnthn** [comment]
@JRaspass Sorry for the delay in looking at this, it seems good to be now. Thanks!

**Pull request merged**